File: 11_动态规划/lis最长上升子序列问题/LISCount.ts
/* eslint-disable no-inner-declarations */

import { bisectLeft } from '../../9_排序和搜索/二分/bisect'
import { sortSearch } from '../../9_排序和搜索/二分/sortSearch'

/** LIS求方案数. */
function countLIS(arr: ArrayLike<number>, strict = true): number {
  const lis: number[][] = []
  const countPreSum: number[][] = []
  for (let i = 0; i < arr.length; i++) {
    const num = arr[i]
    const target = strict ? num : num + 1
    const pos = sortSearch(0, lis.length, i => lis[i][lis[i].length - 1] >= target)
    let count = 1
    if (pos > 0) {
      const tmp1 = lis[pos - 1]
      const k = sortSearch(0, tmp1.length, k => tmp1[k] < target)
      const tmp2 = countPreSum[pos - 1]
      count = tmp2[tmp2.length - 1] - tmp2[k]
    }
    if (pos === lis.length) {
      lis.push([num])
      countPreSum.push([0, count])
    } else {
      lis[pos].push(num)
      const tmp = countPreSum[pos]
      tmp.push(tmp[tmp.length - 1] + count)
    }
  }

  const last = countPreSum[countPreSum.length - 1]
  return last[last.length - 1]
}

const MOD = 1e9 + 7

/** 求长为length的LIS个数. */
function countLISWithLength(arr: ArrayLike<number>, length: number, strict = true): number {
  const copy = Array(arr.length)
  const sorted = Array(arr.length)
  for (let i = 0; i < arr.length; i++) {
    copy[i] = arr[i]
    sorted[i] = arr[i]
  }
  sorted.sort((a, b) => a - b)
  for (let i = 0; i < arr.length; i++) {
    copy[i] = bisectLeft(sorted, arr[i]) + 2
  }

  const n = copy.length
  const add = (tree: number[], i: number, val: number): void => {
    for (; i < n + 2; i += i & -i) {
      tree[i] = (tree[i] + val) % MOD
    }
  }
  const sum = (tree: number[], i: number): number => {
    let res = 0
    for (; i > 0; i &= i - 1) res = (res + tree[i]) % MOD
    return res
  }

  const dp = Array(length + 1)
  for (let i = 0; i <= length; i++) {
    dp[i] = Array(n).fill(0)
  }

  for (let i = 1; i <= length; i++) {
    const tree = Array(n + 2).fill(0)
    const tmp1 = dp[i - 1]
    const tmp2 = dp[i]
    if (i === 1) {
      add(tree, 1, 1)
    }
    if (strict) {
      for (let j = 0; j < n; j++) {
        const v = copy[j]
        tmp2[j] = sum(tree, v - 1)
        add(tree, v, tmp1[j])
      }
    } else {
      for (let j = 0; j < n; j++) {
        const v = copy[j]
        tmp2[j] = sum(tree, v)
        add(tree, v, tmp1[j])
      }
    }
  }

  const last = dp[length]
  let res = 0
  for (let i = 0; i < n; i++) {
    res = (res + last[i]) % MOD
  }
  return res
}

export { countLIS, countLISWithLength }

if (require.main === module) {
  // https://leetcode.cn/problems/number-of-longest-increasing-subsequence/description/
  function findNumberOfLIS(nums: number[]): number {
    return countLIS(nums, true)
  }

  console.log(countLISWithLength([1, 1, 1, 1], 2, false))
}

File: 11_动态规划/lis最长上升子序列问题/LISCount.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

// 求LIS方案数
// LC673 https://leetcode-cn.com/problems/number-of-longest-increasing-subsequence/
// 673. 最长递增子序列的个数
func CountLIS(nums []int, strict bool) int {
	lis := [][]int{}         // 长度为i的lis的结尾的值
	countPreSum := [][]int{} // sum(lis[i][:j])

	for _, v := range nums {
		target := v
		if !strict {
			target++
		}
		pos := sort.Search(len(lis), func(i int) bool { return lis[i][len(lis[i])-1] >= target })
		count := 1
		if pos > 0 {
			i := sort.Search(len(lis[pos-1]), func(i int) bool { return lis[pos-1][i] <= target })
			count = countPreSum[pos-1][len(countPreSum[pos-1])-1] - countPreSum[pos-1][i]
		}
		if pos == len(lis) {
			lis = append(lis, []int{v})
			countPreSum = append(countPreSum, []int{0, count})
		} else {
			lis[pos] = append(lis[pos], v)
			countPreSum[pos] = append(countPreSum[pos], countPreSum[pos][len(countPreSum[pos])-1]+count)
		}
	}
	count := countPreSum[len(countPreSum)-1]
	return count[len(count)-1]
}

const MOD int = 1e9 + 7

// 求长为length的LIS个数.
// 赤壁之战 https://www.acwing.com/problem/content/299/
// 定义 dp[i][j] 表示 a[:j+1] 的长度为 i 且以 a[j] 结尾的 LIS
// 则有 dp[i][j] = ∑dp[i-1][k]  (k<j && a[k]<a[j])
// 注意到当 j 增加 1 时，只多了 k=j 这一个新决策，这样可以用树状数组来维护
// 复杂度 O(n*length*logn)
func CountLISWithLength(nums []int, length int, strict bool) int {
	nums = append(nums[:0:0], nums...)
	_nums := append(nums[:0:0], nums...)
	sort.Ints(_nums)
	for i, v := range nums {
		nums[i] = sort.SearchInts(_nums, v) + 2 // 离散化成从 2 开始的序列
	}

	n := len(nums)
	var tree []int
	add := func(i, val int) {
		for ; i < n+2; i += i & -i {
			tree[i] = (tree[i] + val) % MOD
		}
	}
	sum := func(i int) (res int) {
		for ; i > 0; i &= i - 1 {
			res = (res + tree[i]) % MOD
		}
		return
	}

	dp := make([][]int, length+1)
	for i := range dp {
		dp[i] = make([]int, n)
	}
	for i := 1; i <= length; i++ {
		tree = make([]int, n+2)
		tmp1, tmp2 := dp[i-1], dp[i]
		if i == 1 {
			add(1, 1)
		}
		if strict {
			for j, v := range nums {
				tmp2[j] = sum(v - 1)
				add(v, tmp1[j])
			}
		} else {
			for j, v := range nums {
				tmp2[j] = sum(v)
				add(v, tmp1[j])
			}
		}
	}

	res := 0
	for _, v := range dp[length] {
		res = (res + v) % MOD
	}
	return res
}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var T int
	fmt.Fscan(in, &T)

	for i := 1; i <= T; i++ {
		var n, m int
		fmt.Fscan(in, &n, &m)
		nums := make([]int, n)
		for j := range nums {
			fmt.Fscan(in, &nums[j])
		}
		fmt.Fprintf(out, "Case #%d: %d\n", i, CountLISWithLength(nums, m, true))
	}
}
